llama_index/agent/react/base.py

```diff
@@ -336,27 +364,47 @@ def stream_chat(
             chat_stream = self._llm.stream_chat(input_chat)
 
             # iterate over stream, break out if is final answer after the "Answer: "
-            is_done = False
             full_response = ChatResponse(
                 message=ChatMessage(content=None, role="assistant")
             )
-            for r in chat_stream:
-                if "Answer: " in (r.message.content or ""):
-                    is_done = True
-                    break
-                full_response = r
-            if is_done:
-                break
+            for latest_chunk in chat_stream:
+                latest_content = latest_chunk.message.content
+                if latest_content:
+                    if not latest_content.startswith(
+                        "Thought"
+                    ):  # doesn't follow thought-action format
+                        is_done = True
+                        full_response = latest_chunk
+                        break
+                    else:
+                        if "Answer: " in latest_content:
+                            is_done = True
+                            full_response = latest_chunk
+                            break
+                    full_response = latest_chunk
 
             # given react prompt outputs, call tools or return response
             reasoning_steps, _ = self._process_actions(
-                tools=tools, output=full_response
+                tools=tools, output=full_response, is_streaming=True
             )
             current_reasoning.extend(reasoning_steps)
 
         # Get the response in a separate thread so we can yield the response
+        response_stream = (
+            chain.from_iterable(  # need to add back partial response chunk
+                [
+                    unit_generator(latest_chunk),
+                    chat_stream,
+                ]
+            )
+        )
+        response_stream_c: Generator[ChatResponse, None, None] = cast(
+            Generator[ChatResponse, None, None], response_stream
+        )
```

Comment:
can we pull this into function for readability?

Reply:
absolutely.